libraries/ms-drivers/test/test_voltage_regulator.c

```diff
@@ -0,0 +1,64 @@
+#include "adc.h"
+#include "voltage_regulator.h"
+
+#include "delay.h"
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define TEST_CALLBACK_EXPECTED_TIMES_CALLED 1
+#define TEST_CALLBACK_EXPECTED_CONTEXT 0xA
+
+static const GpioAddress test_enable_pin = { .port = GPIO_PORT_A, .pin = 0 };
+static const GpioAddress test_monitor_pin = { .port = GPIO_PORT_A, .pin = 1 };
+static uint8_t s_regulator_callback_context;
+static uint8_t s_regulator_callback_called;
+
+VoltageRegulatorStorage storage = { 0 };
+
+StatusCode TEST_MOCK(gpio_get_state)(const GpioAddress *address, GpioState *input_state) {
+  *input_state = GPIO_STATE_HIGH;
+  return STATUS_CODE_OK;
+}
+void setup_test(void) {
+  gpio_init();
+  interrupt_init();
+  soft_timer_init();
+  s_regulator_callback_context = 0xA;
+}
+
+static void prv_test_voltage_regulator_callback(void *context, VoltageRegulatorError error) {
+  LOG_DEBUG("VOLTAGE REGULATOR ERROR CALLBACK TRIGGERED\n");
+
+  uint8_t *regulator_context = context;
+  TEST_ASSERT_EQUAL(TEST_CALLBACK_EXPECTED_CONTEXT, *regulator_context);
+
+  s_regulator_callback_called++;
+  TEST_ASSERT_EQUAL(TEST_CALLBACK_EXPECTED_TIMES_CALLED, s_regulator_callback_called);
+}
+
+void test_voltage_regulator_init_works(void) {
+  VoltageRegulatorSettings settings = { .enable_pin = test_enable_pin,
+                                        .monitor_pin = test_monitor_pin,
+                                        .error_callback = &prv_test_voltage_regulator_callback,
+                                        .error_callback_context = &s_regulator_callback_context };
+
+  TEST_ASSERT_OK(voltage_regulator_init(&settings, &storage));
+}
+
+void test_voltage_regulator_set_enabled_works(void) {
+  VoltageRegulatorSettings settings = { .enable_pin = test_enable_pin,
+                                        .monitor_pin = test_monitor_pin,
+                                        .error_callback = &prv_test_voltage_regulator_callback,
+                                        .error_callback_context = &s_regulator_callback_context };
+  voltage_regulator_init(&settings, &storage);
+  TEST_ASSERT_OK(voltage_regulator_set_enabled(&storage, true));
+  delay_s(5);
```

Comment:
I'm not sure what the point of this test is - it's not really checking any of the four scenarios I described on slack.

Instead, I think a better strategy would be to change your `gpio_get_state` mock to, instead of returning `GPIO_STATE_HIGH` all the time, return a static variable, perhaps called `s_gpio_get_return`. Then, for each of the four combinations, set the `gpio_get_state` return value, call `voltage_regulator_set_enabled`, `delay_ms` until the time that the callback should have been called (plus some buffer), and check that it was or was not called as appropriate. Set the configurable polling interval to something short so that the test doesn't take very long - it'll be run on all Travis CI pipelines after this is merged, so we don't want to hold them up by very long.

Reply:
tested all 4 cases. callback was called when 
1. setEnabled(true) and gpiogetstate returns LOW
2. setenabled(false) and gpiogetstate returns HIGH

Reply:
By "test all 4 cases", I meant have 4 tests, each of which test one of the cases, not test them manually by changing one test.

Reply:
done